source/auth/auth.rst

```diff
@@ -1251,172 +1251,527 @@ mechanism
 
 mechanism_properties
     PROVIDER_NAME
-        Drivers MUST allow the user to specify a name for using a service
-        to obtain credentials that is one of ["aws"].
-    REQUEST_TOKEN_CALLBACK
-        Drivers MUST allow the user to specify a callback of the form
-        "onRequest" (defined below), if the driver supports
-        providing objects as mechanism property values.  Otherwise the driver MUST allow it as a MongoClientOption.
-    REFRESH_TOKEN_CALLBACK
-        Drivers MUST allow the user to specify a callback of the form
-        "onRefresh" (defined below), if the driver supports
-        providing objects as mechanism property values.  Otherwise the driver MUST allow it as a MongoClientOption.
+        The name of a built-in OIDC provider integration to use to obtain
+        credentials. The value MUST be one of ["aws"]. If both ``PROVIDER_NAME``
+        and an `OIDC Callback`_ are provided for the same ``MongoClient``
+        (either via the ``CALLBACK`` mechanism property or a ``MongoClient``
+        configuration), the driver MUST raise an error.
+
+    CALLBACK
+        An `OIDC Callback`_ that returns credentials for OIDC providers that do
+        not have a built-in integraiton. Drivers MAY allow the user to specify
+        an `OIDC Callback`_ using a ``MongoClient`` configuration instead of a
+        mechanism property, depending on what is conventional for the driver.
+        Drivers MUST NOT support both the ``CALLBACK`` mechanism property and
+        the ``MongoClient`` configuration.
+
+    CALLBACK_TYPE
```

Comment:
This property seems odd. Consider a MongoClient configuration for the OIDC callback: what should its signature be? For statically typed languages, you would define a type for each callback.  In Java, you might have:

```
interface HumanOidcCallback {
   HumanCallbackReturnType callback(HumanCallbackParams);
}

interface MachineOidcCallback {
   MachineCallbackReturnType callback(MachineCallbackParams);
}

...
   void oidcCallbackType(String type) ...
   void registerOidcCallback(...  What type should this be? In Java the only choice would be Object
```

I think it would be better, at least in Java, to instead overload the second method:

```
   void registerOidcCallback(HumanOidcCallback callback) ...
   void registerOidcCallback(MachineOidcCallback callback) ...
```

and it would be a runtime error to register both, and then the callback type is not necessary.

Also, see my comment below on callback type not being required in all cases, as it's related.

Reply:
In the first example, both callbacks would use the same interface, so there would only be one `registerOidcCallback` method that accepted the `OidcCallback` type. The params and return values for the human auth flow are a superset of the params and return types for the machine auth flow. To support the human auth flow, `OidcCallback` would include all params and return values for both the machine and human auth flows, and `oidcCallbackType` would specify which flow to use.

However, I see your point about the more idiomatic pattern in Java being to use method overriding to set the appropriate OIDC callback. Seems like the pattern for passing callbacks can be very different in different drivers and languages, so I think the current language in the callback API section is probably too prescriptive. I will update that section to be more permissive so drivers can implement a callback API that is idiomatic for their driver and language.